docs(CL): fees spec (#3467)
osmosis-labs/osmosis: docs/architecture/concentrated-liquidity-adr.md

> As a an LP, I want to earn fees on my capital so that I am incentivized to participate in the market making actively.

For our balancer style pools, fees go back directly to the pool to benefit LPs.
For a concentrated liquidity pool, this is no longer possible due to the non-fungible property
of positions. As a result, there is a different accumulator-based mechanism for keeping track
of and storing the fees.

First, note that the fees are collected in tokens themselves rather than in units of liquidity.

> Thus, we need two accumulators for each token. Temporaly, these fee accumulators accessed together

Author: ```suggestion
Thus, we need two accumulators for each token. Temporally, these fee accumulators are accessed together
```

> from state most of the time. Therefore, we define a data structure for storing the fees of each token in the pool.

Author: I meant "temporally", meaning with relation to time. Missed extra L

if the tick being initialized is below the current tick. This is equivalent to assumming that 
all fees have been accrued below the initialized tick.

In the example code snippets below, we only focus on the token0. The token1 is analogous. 

```go
tick.FeeGrowthOutside.Token0 := sdk.ZeroDec()

if initializedTickNum <= pool.CurrentTick {
    tick.FeeGrowthOutside.Token0 = pool.FeeGrowthGlobalOutside.Token0
}
```

Essentially, setting tick's `tick.FeeGrowthOutside.TokenX` to the global `pool.FeeGrowthGlobalOutside.TokenX`
represents the amount of fees collected by the pool up until the tick was activated.

Once a tick is activated again (crossed in either direction), `tick.FeeGrowthOutside.TokenX` is
updated to add the difference between `pool.FeeGrowthGlobalOutsideX` and the old value of
`tick.FeeGrowthOutside.TokenX`. 

```go
tick.FeeGrowthOutside.Token0 =  tick.FeeGrowthOutside.Token0.Add(pool.FeeGrowthGlobalOutside.Token0.Sub(tick.FeeGrowthOutside.Token0))
```

Tracking how much of fees are collected outside of a tick allows us to calculate the amount
of fees inside the position on request.

Intuitively, we update the activated tick with the amount of fees collected for

> every tick lower than the tick that is being been crossed.

Reviewer: ```suggestion
every tick lower than the tick that is being crossed.
```

    feeGrowthBelowLowerTick0 = pool.FeeGrowthOutside.Token0
} else {
    feeGrowthBelowLowerTick0 = pool.FeeGrowthGlobalOutside.Token0 - lowerTick.FeeGrowthOutside.Token0
}
```

We calculate the fee growth above the upper tick in the following way:

```go
var feeGrowthAboveUpperTick0 sdk.Dec

if pool.CurrentTick >= upperTickNum {
    feeGrowthAboveUpperTick0 = pool.FeeGrowthGlobalOutside.Token0 - upperTick.FeeGrowthOutside.Token0
} else {
    feeGrowthAboveUpperTick0 = pool.FeeGrowthOutside.Token0
}
```

Now, by having the fee growth below the lower and above the upper tick of a range,
we can calculate the fee growth inside the range by subtracting the two from the
global per-unit-of-liquidity fee growth.

```go
feeGrowthInsideRange0 := pool.FeeGrowthGlobalOutside.Token0 - feeGrowthBelowLowerTick0 - feeGrowthAboveUpperTick0
```

Note that although `tick.FeeGrowthOutside.Token0` may be initialized at a different
point in time for each tick, the comparison of these values between ticks
is not meaningful. There is also no guarantee that the values
across ticks will follow any particular pattern. 

However, this does not affect the per-position calculations since

> all the position  needs to know is the fee growth inside the position's

Reviewer: ```suggestion
all the position needs to know is the fee growth inside the position's
```

range since the position was last touched.

- **Per-position**

type Position struct {
    FeeGrowthInsideLast Fee
    UncollectedFee Fee
}

> Recall that contrary to traditional pools, in concentrated liquidity pool,

Reviewer: ```suggestion
Recall that contrary to traditional pools, in a concentrated liquidity pool,
```

fees do not get auto re-injected into the pool. Instead, they are tracked by
`position.TokensUncollected0` and `position.TokensUncollected1` fields of each position.

The `position.FeeGrowthInside0Last` and `position.FeeGrowthInside1Last` accumulators
are used to calculate the  _uncollected fees_ to add to `position.TokensUncollected0`
and `position.TokensUncollected1`.

> The amount of uncollected fees needs to be calculated every time a user touches
> the position. That is when a position is created, liqudiity is added or removed.

Reviewer: ```suggestion
The amount of uncollected fees needs to be calculated every time a user modifies
their position. That is when a position is created, liquidity is added or removed.
```

> We must recalculate the values for any modification because with more liqudity

Reviewer: ```suggestion
We must recalculate the values for any modification because with more liquidity
```

added to the position, the amount of fees collected by the position increases.

> Let `feeGrowthInside0` be the amount of fee growth per unit of liqudity within

Reviewer: ```suggestion
Let `feeGrowthInside0` be the amount of fee growth per unit of liquidity within
```

was described in the previous section. Once we have `feeGrowthInside0` computed, we update the
`position.UncollectedFee.Token0` and `position.FeeGrowthInsideLast.Token0`.

```go
// Note that tokensUncollected0 is 0 when a position is created.
if !isPositionNew {
    uncollectefFeeAddition := pool.Liquidity.Mul(feeGrowthInside.Token0.Sub(position.FeeGrowthInsideLast.Token0))
    position.UncollectedFee.Token0 = position.UncollectedFee.Token0.Add(uncollectefFeeAddition)
}

position.FeeGrowthInsideLast.Token0 = feeGrowthInside.Token0
```

##### Collecting Fees

Collecting fees is as simple as transferring the requested amount
from the pool address to the position's owner.

```go
func (k Keeper) collectFees(
    owner sdk.AccAddress,
    lowerTick, upperTick int64, 
    amount0Requested, amount1Requested sdk.Int) error {
    // validate ticks

    // get position if exists

    // make sure that amounts requested are not greater than
    // position's uncollected fees.

    // bank send position.TokensUncollected0 and position.TokensUncollected1
    //from pool address to position owner
    // TODO: revisit to make sure truncations are handled correctly.
}

> ```

Reviewer: Random design choice question: should we give users the ability to specify how much of their share of fees they want to collect (like here) or should we just do an all or nothing collection? IMO, we should allow one collection every 24h (or just reset after epoch), and that collection should be their entire share. Otherwise the META is to just spam collections constantly.

Author: Good point. 

I will change this to:
- auto-collect full amount every epoch
- have the ability to manually collect full amounts (optional, we can implement later)

Reviewer: After our discussion with Dev, I still think this point stands, just not auto-collect but instead allow users to collect once every 24h, so be reset at epoch
